File: NATSInternal.Frontend/NATSInternal.React/src/models/product/productDetailModel.ts
import { createPhotoBasicModel, createProductCategoryBasicModel, createUserBasicModel } from "@/models";
import { getDisplayDateTimeString, getProductUpdateRoutePath, getAmountDisplayText } from "@/helpers";

declare global {
  type ProductDetailModel = Readonly<{
    id: number;
    name: string;
    description: string | null;
    unit: string;
    defaultAmountBeforeVatPerUnit: number;
    defaultVatPercentagePerUnit: number;
    stockingQuantity: number;
    resupplyThresholdQuantity: number;
    isForRetail: boolean;
    isDiscontinued: boolean;
    createdDateTime: string;
    createdUser: UserBasicModel;
    lastUpdatedDateTime: string | null;
    lastUpdatedUser: UserBasicModel | null;
    deletedUser: UserBasicModel | null;
    deletedDateTime: string | null;
    categories: ProductCategoryBasicModel[];
    photos: PhotoBasicModel[];
    authorization: ProductExistingAuthorizationResponseDto;
    formattedDefaultAmountBeforeVatPerUnit: string;
    updateRoutePath: string;
  }>;
}

export function createProductDetailModel(responseDto: ProductDetailResponseDto): ProductDetailModel {
  return {
    id: responseDto.id,
    name: responseDto.name,
    description: responseDto.description,
    unit: responseDto.unit,
    defaultAmountBeforeVatPerUnit: responseDto.defaultAmountBeforeVatPerUnit,
    defaultVatPercentagePerUnit: responseDto.defaultVatPercentagePerUnit,
    stockingQuantity: responseDto.stockingQuantity,
    resupplyThresholdQuantity: responseDto.resupplyThresholdQuantity,
    isForRetail: responseDto.isForRetail,
    isDiscontinued: responseDto.isDiscontinued,
    createdDateTime: getDisplayDateTimeString(responseDto.createdDateTime),
    createdUser: createUserBasicModel(responseDto.createdUser),
    lastUpdatedDateTime: responseDto.lastUpdatedDateTime &&
      getDisplayDateTimeString(responseDto.lastUpdatedDateTime),
    lastUpdatedUser: responseDto.lastUpdatedUser && createUserBasicModel(responseDto.lastUpdatedUser),
    deletedDateTime: responseDto.deletedDateTime && getDisplayDateTimeString(responseDto.deletedDateTime),
    deletedUser: responseDto.deletedUser && createUserBasicModel(responseDto.deletedUser),
    categories: responseDto.categories.map(createProductCategoryBasicModel),
    photos: responseDto.photos.map(dto => createPhotoBasicModel(dto)),
    formattedDefaultAmountBeforeVatPerUnit: getAmountDisplayText(responseDto.defaultAmountBeforeVatPerUnit),
    updateRoutePath: getProductUpdateRoutePath(responseDto.id),
    authorization: responseDto.authorization
  };
}

File: NATSInternal.Web/Models/Product/ProductDetailModel.cs
using NATSInternal.Application.UseCases.Products;
using NATSInternal.Web.Models.Shared;

namespace NATSInternal.Web.Models;

public class ProductDetailModel
{
    #region Constructors
    public ProductDetailModel(ProductGetDetailResponseDto responseDto)
    {
        Id = responseDto.Id;
        Name = responseDto.Name;
        Description = responseDto.Description;
        Unit = responseDto.Unit;
        DefaultAmountBeforeVatPerUnit = responseDto.DefaultAmountBeforeVatPerUnit;
        DefaultVatPercentagePerUnit = responseDto.DefaultVatPercentagePerUnit;
        StockingQuantity = responseDto.StockingQuantity;
        IsForRetail = responseDto.IsForRetail;
        IsDiscontinued = responseDto.IsDiscontinued;
        CreatedDateTime = responseDto.CreatedDateTime;
        CreatedUser = new(responseDto.CreatedUser);
        LastUpdatedDateTime = responseDto.LastUpdatedDateTime;
        Photos = responseDto.Photos.Select(p => new PhotoBasicModel(p)).ToList().AsReadOnly();
        Authorization = new(responseDto.Authorization);

        if (responseDto.LastUpdatedDateTime.HasValue)
        {
            LastUpdatedDateTime = responseDto.LastUpdatedDateTime;
        }

        if (responseDto.LastUpdatedUser is not null)
        {
            LastUpdatedUser = new(responseDto.LastUpdatedUser);
        }

        if (responseDto.Brand is not null)
        {
            Brand = new(responseDto.Brand);
        }

        if (responseDto.Category is not null)
        {
            Category = new(responseDto.Category);
        }
    }
    #endregion

    #region Properties
    public Guid Id { get; }
    public string Name { get; }
    public string? Description { get; }
    public string Unit { get; }
    public long DefaultAmountBeforeVatPerUnit { get; }
    public int DefaultVatPercentagePerUnit { get; }
    public int StockingQuantity { get; }
    public bool IsForRetail { get; }
    public bool IsDiscontinued { get; }
    public DateTime CreatedDateTime { get; }
    public UserBasicModel CreatedUser { get; }
    public DateTime? LastUpdatedDateTime { get; }
    public UserBasicModel? LastUpdatedUser { get; }
    public BrandBasicModel? Brand { get; }
    public ProductCategoryBasicModel? Category { get; }
    public IReadOnlyList<PhotoBasicModel> Photos { get; }
    public ProductExistingAuthorizationModel Authorization { get; }
    #endregion
}